// elfinder-node/dist/lfs.utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleChunkUpload = exports.volume = exports.suffix = exports.readdir = exports.parse = exports.move = exports.init = exports.info = exports.extract = exports.encode = exports.decode = exports.copy = exports.compress = void 0;
const lz = __importStar(require("lzutf8"));
const path = __importStar(require("path"));
const mime = __importStar(require("mime-types"));
const _ = __importStar(require("underscore"));
const fs = __importStar(require("fs-extra"));
const archiver = __importStar(require("archiver"));
const adm_zip_1 = __importDefault(require("adm-zip"));
const util_1 = require("util");
const compress = async (files, dest, config) => {
    return new Promise((resolve, reject) => {
        const output = fs.createWriteStream(dest);
        const archive = archiver.default('zip', {
            store: true,
        });
        output.on('close', () => {
            resolve(true);
        });
        archive.on('error', (err) => {
            console.log(err);
            reject(err);
        });
        archive.pipe(output);
        _.each(files, (file) => {
            const target = (0, exports.decode)(file, config);
            if (fs.lstatSync(target.absolutePath).isDirectory()) {
                const name = path.basename(target.absolutePath);
                archive.directory(path.normalize(target.absolutePath + path.sep), name);
            }
            else {
                archive.file(target.absolutePath, {
                    name: target.name,
                });
            }
        });
        archive.finalize();
    });
};
exports.compress = compress;
const copy = async (opts, config) => {
    const fileExists = await fs.pathExists(opts.dst);
    if (fileExists)
        throw new Error('Destination exists');
    await fs.copy(opts.src, opts.dst);
    const infoResult = await (0, exports.info)(opts.dst, config);
    return {
        added: [infoResult],
        changed: [(0, exports.encode)(path.dirname(opts.dst), config)],
    };
};
exports.copy = copy;
const decode = (dir, config) => {
    let root, name, volume;
    if (!dir || dir.length < 4)
        throw Error('Invalid Path');
    if (dir[0] !== 'v' || dir[2] !== '_')
        throw Error('Invalid Path');
    volume = parseInt(dir[1]);
    let relative = dir
        .substr(3, dir.length - 3)
        .replace(/-/g, '+')
        .replace(/_/g, '/')
        .replace(/\./g, '=');
    relative = lz.decompress(relative + '==', {
        inputEncoding: 'Base64',
    });
    name = path.basename(relative);
    root = config.path;
    return {
        volume,
        dir: root,
        path: relative,
        name,
        absolutePath: path.join(root, relative),
    };
};
exports.decode = decode;
const encode = (dir, config) => {
    const parsedInfo = (0, exports.parse)(dir, config);
    const relative = lz
        .compress(parsedInfo.path, {
        outputEncoding: 'Base64',
    })
        .replace(/=+$/g, '')
        .replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=/g, '.');
    return 'v' + parsedInfo.volume + '_' + relative;
};
exports.encode = encode;
const extract = async (source, dest) => {
    const zip = new adm_zip_1.default(source);
    const files = zip.getEntries().map((file) => file.entryName);
    const extractAsync = (0, util_1.promisify)(zip.extractAllToAsync.bind(zip));
    await extractAsync(dest, true);
    return files;
};
exports.extract = extract;
const info = async (p, config) => {
    const parsedInfo = (0, exports.parse)(p, config);
    if (parsedInfo.volume < 0)
        throw new Error('Volume not found');
    const stat = await fs.stat(p);
    const r = {
        name: path.basename(p),
        size: stat.size,
        hash: (0, exports.encode)(p, config),
        mime: stat.isDirectory()
            ? 'directory'
            : mime.lookup(p) || 'application/binary',
        ts: Math.floor(stat.mtime.getTime() / 1000),
        volumeid: 'v' + parsedInfo.volume + '_',
        read: 1,
        write: 1,
        locked: 0,
        isdir: false,
    };
    if (r.mime.indexOf('image/') === 0) {
        const filename = (0, exports.encode)(p, config);
        console.log('bbb', config.tmbroot, filename);
        const tmbPath = path.join(config.tmbroot, filename + '.png');
        if (await fs.pathExists(tmbPath)) {
            r.tmb = filename + '.png';
        }
        else {
            r.tmb = '1';
        }
    }
    if (!parsedInfo.isRoot) {
        const parent = path.dirname(p);
        r.phash = (0, exports.encode)(parent, config);
    }
    else {
        r.options = {
            disabled: config.disabled,
            archivers: {
                create: ['application/zip'],
                extract: ['application/zip'],
                createext: {
                    'application/zip': 'zip',
                },
            },
            url: config.URL,
        };
        if (config.icon) {
            r.options.csscls = config.icon;
        }
    }
    const acl = config.acl(p);
    r.read = acl.read;
    r.write = acl.write;
    r.locked = acl.locked;
    r.isdir = r.mime === 'directory';
    if (r.isdir) {
        const items = await fs.readdir(p);
        for (const item of items) {
            if ((await fs.lstat(path.join(p, item))).isDirectory()) {
                r.dirs = 1;
                break;
            }
        }
    }
    return r;
};
exports.info = info;
const init = async (config) => {
    const volumeInfo = await (0, exports.info)(config.path, config);
    volumeInfo.phash = '';
    return [volumeInfo];
};
exports.init = init;
const move = async (opts, config) => {
    if (await fs.pathExists(opts.dst)) {
        throw new Error('Destination exists');
    }
    await fs.move(opts.src, opts.dst);
    const infoResult = await (0, exports.info)(opts.dst, config);
    return {
        added: [infoResult],
        removed: opts.upload ? [] : [(0, exports.encode)(opts.src, config)],
    };
};
exports.move = move;
const parse = (p, config) => {
    const v = (0, exports.volume)(p, config);
    const root = config.path || '';
    let relative = p.substr(root.length, p.length - root.length);
    if (relative.indexOf(path.sep) !== 0)
        relative = path.sep + relative;
    return {
        volume: v,
        dir: root,
        path: relative,
        isRoot: relative === path.sep,
    };
};
exports.parse = parse;
const readdir = async (dir) => {
    const items = await fs.readdir(dir);
    const files = [];
    for (const item of items) {
        const itemInfo = await fs.lstat(path.join(dir, item));
        files.push({
            name: item,
            isdir: itemInfo.isDirectory(),
        });
    }
    return files;
};
exports.readdir = readdir;
const suffix = (name, suff) => {
    const ext = path.extname(name);
    const fil = path.basename(name, ext);
    return fil + suff + ext;
};
exports.suffix = suffix;
const volume = (p, config) => {
    // For single volume, always return 0 if path matches
    if (p.indexOf(config.path) === 0) {
        return 0;
    }
    return -1;
};
exports.volume = volume;
/**
 * Handles uploading a single chunk of a file
 * Returns information about whether the upload is complete and where the file is located
 */
const handleChunkUpload = async (opts) => {
    const { chunkName, chunkFile, range, destinationDir } = opts;
    const [start, , totalSize] = range.split(',').map(Number);
    // Extract real filename by removing the chunk pattern (.N_M.part)
    const realFilename = chunkName.replace(/\.\d+_\d+\.part$/, '');
    const finalPath = path.join(destinationDir, realFilename);
    const chunkPath = path.join(destinationDir, chunkName);
    // Ensure destination directory exists
    await fs.ensureDir(destinationDir);
    // Move uploaded chunk to temp location
    await fs.move(chunkFile, chunkPath, { overwrite: true });
    // Check if this is the last chunk by checking file size
    const chunkStat = await fs.stat(chunkPath);
    const isLastChunk = start + chunkStat.size >= totalSize;
    if (isLastChunk) {
        // Merge all chunks into final file
        await mergeChunks(realFilename, destinationDir);
        return { isComplete: true, finalPath, chunkPath };
    }
    return { isComplete: false, chunkPath };
};
exports.handleChunkUpload = handleChunkUpload;
/**
 * Merges all chunk files for a given filename into a single file
 */
const mergeChunks = async (filename, directory) => {
    const escapedFilename = filename.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    const chunkPattern = new RegExp(`^${escapedFilename}\\.\\d+_\\d+\\.part$`);
    // Find and sort all chunk files
    const chunkFiles = (await fs.readdir(directory))
        .filter((f) => chunkPattern.test(f))
        .sort((a, b) => {
        // Extract chunk numbers for proper sorting
        const aMatch = a.match(/\.(\d+)_\d+\.part$/);
        const bMatch = b.match(/\.(\d+)_\d+\.part$/);
        const aNum = aMatch ? parseInt(aMatch[1]) : 0;
        const bNum = bMatch ? parseInt(bMatch[1]) : 0;
        return aNum - bNum;
    });
    const finalPath = path.join(directory, filename);
    const writeStream = fs.createWriteStream(finalPath);
    // Write each chunk in order and clean up
    for (const chunk of chunkFiles) {
        const chunkFullPath = path.join(directory, chunk);
        const data = await fs.readFile(chunkFullPath);
        writeStream.write(data);
        await fs.remove(chunkFullPath);
    }
    // Finalize the write stream
    await new Promise((resolve, reject) => {
        writeStream.end((err) => {
            if (err)
                reject(err);
            else
                resolve();
        });
    });
};
